src/main/scala/analysis/DSAUtility.scala

```diff
@@ -0,0 +1,983 @@
+package analysis
+
+import analysis.solvers.{DSAUnionFindSolver, UnionFindSolver, Var}
+import cfg_visualiser.{DotStruct, DotStructElement, StructArrow, StructDotGraph}
+import ir.{Assign, BVADD, BinaryExpr, BitVecLiteral, BitVecType, CFGPosition, DirectCall, Expr, Extract, IntraProcIRCursor, Literal, Memory, MemoryAssign, MemoryLoad, Procedure, Register, Repeat, SignExtend, UnaryExpr, Variable, ZeroExtend, begin, computeDomain, toShortString}
+import specification.{ExternalFunction, SpecGlobal, SymbolTableEntry}
+
+import scala.util.control.Breaks.{break, breakable}
+import scala.collection.mutable
+
+object NodeCounter {
+  var counter: Int = 0
+
+  def getCounter: Int =
+    counter = counter + 1
+    counter
+}
+
+/**
+ * Data Structure Graph for DSA
+ * @param proc procedure of DSG
+ * @param constProp
+ * @param varToSym mapping flow-sensitive (position sensitive) mapping from registers to their set of symbolic accesses
+ * @param globals
+ * @param globalOffsets
+ * @param externalFunctions
+ * @param reachingDefs
+ * @param writesTo
+ * @param params
+ */
+class DSG(val proc: Procedure,
+          constProp: Map[CFGPosition, Map[Variable, FlatElement[BitVecLiteral]]],
+          varToSym: Map[CFGPosition, Map[Variable, Set[SymbolicAccess]]],
+          globals: Set[SymbolTableEntry], globalOffsets: Map[BigInt, BigInt],
+          externalFunctions: Set[ExternalFunction],
+          val reachingDefs: Map[CFGPosition, Map[Variable, Set[CFGPosition]]],
+          val writesTo: Map[Procedure, Set[Register]],
+          val params: Map[Procedure, Set[Variable]]
+         ) {
+
+  // DSNodes owned by this graph, only updated once analysis is done,
+  val nodes: mutable.Set[DSN] = mutable.Set()
+
+
+  // represent callees in proc
+  val callsites: mutable.Set[CallSite] = mutable.Set()
+
+  val mallocRegister = Register("R0", 64)
+  val stackPointer = Register("R31", 64)
+
+  // this is the mapping from offsets/positions on the stack to their representative DS nodes
+  val stackMapping: mutable.Map[BigInt, DSN] =
+    computeDomain(IntraProcIRCursor, Set(proc)).toSeq.sortBy(_.toShortString).foldLeft(Map[BigInt, DSN]()) {
+      (results, pos) => stackBuilder(pos, results)
+    }.to(collection.mutable.Map)
+
+
+  /**
+   * this function takes a stackMapping and updates it based on a memory load and memory store
+   * @param pos memory load or store IL position
+   * @param index memory location of load or store
+   * @param size size of the load or store
+   * @param m stack mapping
+   * @return updated stack mapping
+   */
+  private def visitStackAccess(pos: CFGPosition, index: Expr, size: Int, m: Map[BigInt, DSN]) : Map[BigInt, DSN] =
+    val byteSize = (size.toDouble / 8).ceil.toInt
```

Comment:
what is the point of converting to bytes like this, and why bother casting to double and back to int? A memory access size must always be a multiple of 8 bits.

Reply:
addressing is done based on bytes, so to be consistent access sizes are converted to bytes to determine overlapping accesses based on the address of the accesses and their known size. the casting is unnecessary and will be removed.